handler/src/test/java/io/netty/handler/ssl/CloseNotifyTest.java

```diff
@@ -0,0 +1,235 @@
+            // respond with data and close_notify:
+            serverChannel.writeOutbound(writeAscii(ALLOC, "response_msg"));
+            assertThat(serverChannel.finish(), is(true));
+            assertThat(serverEventQueue.poll(), instanceOf(SslCloseCompletionEvent.class));
+            assertThat(clientEventQueue, empty());
+
+            // consume server response with close_notify:
+            forwardAllWithCloseNotify(serverChannel, clientChannel);
+            assertThat(clientEventQueue.poll(), equalTo((Object) writeAscii(ALLOC, "response_msg")));
+            assertThat(clientEventQueue.poll(), instanceOf(SslCloseCompletionEvent.class));
+
+            // make sure client automatically responds with close_notify:
+            if (!jdkTls13()) {
+                // JDK impl of TLSv1.3 does not automatically generate "close_notify" in response to the received
+                // "close_notify" alert. This is a legit behavior according to the spec:
+                // https://tools.ietf.org/html/rfc8446#section-6.1. Handle it differently:
+                assertCloseNotify((ByteBuf) clientChannel.readOutbound());
+            }
+        } finally {
+            try {
+                clientChannel.finish();
+            } finally {
+                serverChannel.finish();
+            }
+        }
+
+        if (jdkTls13()) {
+            assertCloseNotify((ByteBuf) clientChannel.readOutbound());
+        } else {
+            discardEmptyOutboundBuffers(clientChannel);
+        }
+
+        assertThat(clientEventQueue.poll(), is(INACTIVE));
+        assertThat(clientEventQueue, empty());
+        assertThat(serverEventQueue.poll(), is(INACTIVE));
+        assertThat(serverEventQueue, empty());
+
+        assertThat(clientChannel.releaseInbound(), is(false));
+        assertThat(clientChannel.releaseOutbound(), is(false));
+        assertThat(serverChannel.releaseInbound(), is(false));
+        assertThat(serverChannel.releaseOutbound(), is(false));
+    }
+
+    private boolean jdkTls13() {
+        return provider == SslProvider.JDK && PROTOCOL_TLS_V1_3.equals(protocol);
+    }
+
+    private EmbeddedChannel initChannel(final boolean useClientMode,
+            final BlockingQueue<Object> eventQueue) throws Exception {
+
+        SelfSignedCertificate ssc = new SelfSignedCertificate();
+        final SslContext sslContext = (useClientMode
+                ? SslContextBuilder.forClient().trustManager(InsecureTrustManagerFactory.INSTANCE)
+                : SslContextBuilder.forServer(ssc.certificate(), ssc.privateKey()))
+                 .sslProvider(provider)
+                 .protocols(protocol)
+                .build();
+        final EmbeddedChannel channel = new EmbeddedChannel(
+                // use sslContext.newHandler(ALLOC) instead of new SslHandler(sslContext.newEngine(ALLOC)) to create
+                // non-JDK compatible OpenSSL engine that can process partial packets:
+                sslContext.newHandler(ALLOC),
```

Comment:
@Scottmitch I found this confusing that the behavior is different for `ReferenceCountedOpenSslContext` based on how you create the engine. Sure, `newHandler` has a note in javadoc, but `newEngine` doesn't and it's hard to realize that there is another way to build `SslHandler`. We should enhance javadoc. Ideally, we should have `SslContext#newEngine` overload that takes `jdkCompatibilityMode` boolean. That way discoverability will be way better.

https://github.com/netty/netty/commit/f7b3caeddc5bb1da75aaafa4a66dec88ed585d69

Reply:
Since `jdkCompatibilityMode=true` is an optimization which violates the `SSLEngine` API but `SslHandler` can compensate for behavior difference. The initial motivation was to avoid explicitly exposing it on the public API and just enable it internally if the engine is known to be used with `SslHandler`. If we added a `newEngine(boolean jdkCompatibilityMode)` overload there is a chance folks may use the `SSLEngine` outside the `SslHandler`. We could add javadocs on `SslContext#newEngine` highlighting we may apply optimizations applied if you use the `newHandler` then `engine()` methods.